staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -484,3 +490,192 @@ func TestCacheWatcherStoppedInAnotherGoroutine(t *testing.T) {
 		w.Stop()
 	}
 }
+
+func TestTimeBucketWatchersBasic(t *testing.T) {
+	filter := func(_ string, _ labels.Set, field fields.Set) bool {
+		return true
+	}
+	forget := func() {}
+
+	newWatcher := func(deadline time.Time) *cacheWatcher {
+		return newCacheWatcher(0, filter, forget, testVersioner{}, deadline, true)
+	}
+
+	clock := clock.NewFakeClock(time.Now())
+	watchers := newTimeBucketWatchers(clock)
+	now := clock.Now()
+	watchers.addWatcher(newWatcher(now.Add(10 * time.Second)))
+	watchers.addWatcher(newWatcher(now.Add(20 * time.Second)))
+	watchers.addWatcher(newWatcher(now.Add(20 * time.Second)))
+
+	if len(watchers.watchersQueue) != 2 {
+		t.Errorf("unexpected bucket size: %#v", watchers.watchersQueue)
+	}
+	watchers0 := watchers.popExpiredWatchers()
+	if len(watchers0) != 0 {
+		t.Errorf("unexpected bucket size: %#v", watchers0)
+	}
+
+	clock.Step(10 * time.Second)
+	watchers1 := watchers.popExpiredWatchers()
+	if len(watchers1) != 1 || len(watchers1[0]) != 1 {
+		t.Errorf("unexpected bucket size: %v", watchers1)
+	}
+	watchers1 = watchers.popExpiredWatchers()
+	if len(watchers1) != 0 {
+		t.Errorf("unexpected bucket size: %#v", watchers1)
+	}
+
+	clock.Step(12 * time.Second)
+	watchers2 := watchers.popExpiredWatchers()
+	if len(watchers2) != 1 || len(watchers2[0]) != 2 {
+		t.Errorf("unexpected bucket size: %#v", watchers2)
+	}
+}
+
+func testCacherSendBookmarkEvents(t *testing.T, watchCacheEnabled, allowWatchBookmarks, expectedBookmarks bool) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.WatchBookmark, watchCacheEnabled)()
+	backingStorage := &dummyStorage{}
+	cacher, _ := newTestCacher(backingStorage, 1000)
+	defer cacher.Stop()
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+	pred := storage.Everything
+	pred.AllowWatchBookmarks = allowWatchBookmarks
+
+	w, err := cacher.Watch(context.TODO(), "pods/ns", "0", pred)
```

Comment:
Don't you need to set deadline here?

Reply:
It will be add to the first bucket if not set, but set one will be better to understand. done